src/audio/kpb.c

```diff
@@ -1626,6 +1692,51 @@ static void kpb_drain_samples(void *source, struct audio_stream __sparse_cache *
 	}
 }
 
+#if KPB_HIFI4
+static void kpb_convert_32b_to_24b(const struct audio_stream __sparse_cache *source, int ioffset,
+				   void *linear_sink, int ooffset, unsigned int n_samples)
+{
+	int ssize = audio_stream_sample_bytes(source); /* src fmt == sink fmt */
+	uint8_t *in = audio_stream_wrap(source, (uint8_t *)source->r_ptr + ioffset * ssize);
+	uint8_t *out = (uint8_t *)linear_sink + ooffset * ssize;
+	const ae_int32 *in_ptr = (const ae_int32 *)(in);
+	ae_valign align_out = AE_ZALIGN64();
+
+	for (size_t i = 0; i < n_samples; i++) {
+		ae_int32 d32 = in_ptr[i] >> 8;
+		ae_int24 d24 = AE_MOVINT24_FROMINT32(d32);
+
+		AE_SA24_L_IP(d24, align_out, out);
+	}
+	AE_SA64POS_FP(align_out, out);
+}
+#else
+static void kpb_convert_32b_to_24b(const struct audio_stream __sparse_cache *source, int ioffset,
+				   void *sink, int ooffset, unsigned int samples)
+{
+	int ssize = audio_stream_sample_bytes(source);
+	int32_t *src = audio_stream_wrap(source, (int32_t *)source->r_ptr + ioffset * ssize);
+	uint8_t *dst = (uint8_t *)sink + ooffset * 3;
+	int processed;
+	int nmax, i, n;
+
+	for (processed = 0; processed < samples; processed += n) {
+		src = audio_stream_wrap(source, src);
+		n = samples - processed;
+		nmax = KPB_BYTES_TO_S32_SAMPLES(audio_stream_bytes_without_wrap(source, src));
+		n = MIN(n, nmax);
+		for (i = 0; i < n; i += 1) {
+			*dst = *src & 0xFF;
+			dst++;
+			*dst = (*src >> 8) & 0xFF;
+			dst++;
+			*dst = (*src >> 16) & 0xFF;
+			dst++;
```

Comment:
I would write it as
```
	dst[0] = *src & 0xFF;
	dst[1] = (*src >> 8) & 0xFF;
	dst[2] = (*src >> 16) & 0xFF;
	dst += 3;
```
but I'm not sure if the compiler would generate better code from this